### ISSUE TITLE
Add ability to pause and resume a subscription [Stripe]

### ISSUE THREAD

**Issue by tnorthcutt** [restrictcontentpro/restrict-content-pro#1748]
_This is an offshoot of #508; opening an issue specifically for Stripe since we'll be supporting this on a per-gateway basis._

Stripe suggests [four methods for pausing a subscription](https://stripe.com/docs/subscriptions/canceling-pausing#pausing-a-subscription):

> Stripe’s subscriptions logic does not support true pausing of a subscription. However, there are ways to emulate this behavior:
> 1. To “pause” a subscription for a period and then automatically resume the subscription, use a trial period to change the billing cycle. You may or may not want to prorate the move, depending upon your business model.
> 2. To keep the subscription effectively active but no longer billing the customer, apply a 100% off coupon that lasts forever. Later remove the coupon when it’s time to begin billing again.
> 3. Similarly, you could move the customer to a free plan and then resubscribe them to the paid plan later. This approach allows you to easily track paused subscriptions.
> 4. Finally, the cleanest option is to cancel the current subscription and recreate it when it’s time to bill the customer again. This is the least hack-y of the solutions, but the customer doesn’t show as having a subscription during the “pause”.

A few things we'll need to sort out, which will affect which method we choose:

1. Should members retain access while their subscription is paused?
2. Should members be able to pause their own subscriptions, or should this be an admin-only feature?
3. If `#1` is "no", should members retain access to their account settings/billing/etc. while paused?

I'm sure there are more; please comment and I'll update this list with what others have to add.

My thoughts:

Method 4 (canceling/recreating) is a non-starter, IMO.

No, members should not retain access to restricted content while paused. To achieve this, we'll either need to maintain knowledge of a `paused` status (though keep in mind that doing away with statuses entirely has been discussed recently, IIRC, because of the move to support for multiple subscriptions), or use method 3 above, where a member is moved to a free plan for the duration of the pause.

Ideally members would be able to pause their own subscription, but IMO launching this feature as admin-only is fine, and maybe even better. I suspect we'd be able to learn some from how people use the feature in the real world, and make improvements before it's made end-user-facing.

IMO members should retain access to their account settings/dashboard area, and by default a message (configurable?) should appear there (or perhaps on every page when logged in) informing/reminding them that their subscription is paused, and to contact the site owner/admin to un-pause it (or directions to do so themselves, if un-pausing is something a user can do themselves).

**Comment by pippinsplugins** [comment]
1. I'd think members would retain access up to the expiration date set by their most recent payment. Once that is past, access would be revoked but the membership would _not_ be expired.

2. If enabled by an admin opt-in, yes I think so.

3. Yes, members should have access to their settings and billing.

**Comment by nosegraze** [comment]
I think this could be nicely integrated into a slight refactoring of how auto renew is presented. Right now we kind of present two different types of subscription levels:

1. You have an auto recurring subscription.
2. You don't have an auto recurring subscription.

We probably have this distinction because code-wise they're very different (set up a subscription vs process a one-time transaction). But to the end user, there really shouldn't be such a big difference between the two.

If you want to move from recurring to non-recurring, you have to cancel your subscription (this isn't so bad). But if you want to move from non-recurring to recurring, you have to completely re-register. This is the step that's kind of a pain and we can improve on. Instead of presenting these as two separate options, I'd love to see recurring as more of a toggle. You can toggle it off and on at will. I imagine toggling off auto renew is kind of the same as pausing a subscription. I imagine it working much how hosting and domains work where you can just toggle auto renew off and on without needing to re-register or anything like that.

**Comment by tnorthcutt** [comment]
@nosegraze I assume what you're describing (toggling auto-renew on and off at will) would be at the discretion of the site owner. Is that accurate?

Meaning, a site owner could control whether auto-renew is user-controllable or not from the RCP settings.

**Comment by nosegraze** [comment]
I think we could tie this to the existing auto renew setting. We have three options and they could definitely be applied here:

1. Never auto renew - cannot sign up for auto renew and cannot toggle auto renew after the fact.
2. Always auto renew - all signups are auto renew. You cannot toggle auto renew after the fact - only cancel.
3. Let the customer decide - customers can decide via a checkbox at registration if they want auto renew or not. They can then also toggle auto renew after the fact.

So basically only that third option would get the toggle.

**Comment by tnorthcutt** [comment]
@nosegraze makes sense.

Do  you think it's worth expanding this to make it a per-level setting (possibly defaulting to whatever is set in general settings)?

**Comment by nosegraze** [comment]
That should wait for https://github.com/restrictcontentpro/restrict-content-pro/issues/548
